src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/Interfaces.cs

```diff
@@ -29,7 +30,7 @@ public static class ShapeProviderExtensions
     {
         public static IServiceCollection AddShapeAttributes<T>(this IServiceCollection services) where T : class, IShapeAttributeProvider
         {
-            services.AddScoped<T>();
+            services.TryAddScoped<T>();
             services.AddScoped<IShapeAttributeProvider>(sp => sp.GetService<T>());
```

Comment:
Normally in a given module we don't call twice `AddShapeAttributes()` for a given shape type.

Otherwise we would need to also use `TryAddScoped()` here but it may not work when we pass a factory delegate.

Maybe we could have our own `TryAddScoped()` that would be used before line 33 and that would return a bool, so that if it returns false we would not re-register this `IShapeAttributeProvider` too.

Reply:
Technically it's not issue as it's same class registered more than once and only last one is used. However registration is needed only once.

Useful when having shared shape library  not attached to any module , that adds such shapes to service collection, and can be safely called more than once from different modules

Reply:
@ns8482e 

What I meant is that we could also prevent to re-register

    services.AddScoped<IShapeAttributeProvider>(sp => sp.GetService<T>());

But `TryAddScoped()` only checks the service type, here `IShapeAttributeProvider`;

So what you could use is `TryAddEnumerable()`, was not sure so I looked at the dotnet code (see below), because it also checks the implementation type, anf if it is an implementation factory as here, it also checks the return type of the factory.

    services.TryAddEnumerable(
        ServiceDescriptor.Scope<IShapeAttributeProvider>(sp => sp.GetRequiredService<T>()));

https://github.com/dotnet/runtime/blob/231255dac0fe015ebc6fb55466c75cf5b4e05f56/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs#L451

https://github.com/dotnet/runtime/blob/231255dac0fe015ebc6fb55466c75cf5b4e05f56/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceDescriptor.cs#L104

Reply:
Updated to use `TryAddEnumerable`